### PR TITLE
ERC-721 extension

### PR DESCRIPTION
**Pull request by damianmarti**
An ERC-721 extension to demonstrate how to integrate an ERC-721 NFT.

localhost_3000_erc721

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,30 +1,27 @@
-# 🔌 create-eth Extensions
+# ERC-721 NFT Extension for Scaffold-ETH 2
 
-This repository holds all the BG curated extensions for [create-eth](https://github.com/scaffold-eth/create-eth), so you can extend the functionality of your Scaffold-ETH project.
+This extension introduces an ERC-721 token contract and demonstrates how to use it, including getting the total supply and holder balance, listing all NFTs and NFTs from an address, and how to transfer.
 
-## Usage
+The ERC-721 Token Standard introduces a standard for Non-Fungible Tokens ([EIP-721](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol)), in other words, each Token is unique.
```
Comment (Pabl0cks) [inline README.md:5]: `each Token is unique` I think `token` can be uncapitalized there

File: README.md [hunk 2]
```diff
@@ -1,30 +1,27 @@
...
+
+The ERC-721 token contract is implemented using the [ERC-721 token implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol) from OpenZeppelin.
+
+The ERC-721 token implementation uses the [ERC-721 Enumerable extension](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/extensions/ERC721Enumerable.sol) from OpenZeppelin to list all tokens and all the tokens owned by an address. You can remove this if you plan to use an indexer, like a Subgraph or Ponder (extensions available).
```
Comment (Pabl0cks) [inline README.md:9]: Maybe we could add the link to the extension list from se website: ([extensions available](https://scaffoldeth.io/extensions))

Comment (Pabl0cks) [inline README.md:9]: Maybe add again `from the contract` or `from the collection` to the sentence `from OpenZeppelin to list all tokens`

File: README.md [hunk 3]
```diff
@@ -1,30 +1,27 @@
...
+
+## Installation
 
 You can install any of the extensions in this repository by running the following command:
```
Comment (Pabl0cks) [inline README.md:13]: I think we can get rid of this sentence, looks generic and i think the command is enough to understand

File: extension/packages/nextjs/app/erc721/components/AllNfts.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
+"use client";
+
+import { useEffect, useState } from "react";
+import { NFTCard } from "./NFTCard";
+import { useAccount } from "wagmi";
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/AllNfts.tsx:5]: We can delete that

File: extension/packages/nextjs/app/erc721/components/AllNfts.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+export interface Collectible {
+  id: number;
+  uri: string;
+  owner: string;
+  image: string;
+  name: string;
+}
+
+export const AllNfts = () => {
+  const [myNfts, setMyNfts] = useState<Collectible[]>([]);
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/AllNfts.tsx:18]: allNfts, setAllNfts

File: extension/packages/nextjs/app/erc721/components/AllNfts.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  });
+
+  const { data: totalSupply } = useScaffoldReadContract({
+    contractName: "SE2NFT",
+    functionName: "totalSupply",
+    watch: true,
+  });
+
+  useEffect(() => {
+    const updateMyNfts = async (): Promise<void> => {
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/AllNfts.tsx:32]: updateAllNfts

File: extension/packages/nextjs/app/erc721/components/AllNfts.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+          collectibleUpdate.push({
+            id: parseInt(tokenId.toString()),
+            uri: tokenURI,
+            owner,
+            image: metadata.image,
+            name: metadata.name,
+          });
+        } catch (e) {
+          notification.error("Error fetching your NTFs");
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/AllNfts.tsx:56]: "Error fetching NFTs"

File: extension/packages/nextjs/app/erc721/components/MyNfts.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+          collectibleUpdate.push({
+            id: parseInt(tokenId.toString()),
+            uri: tokenURI,
+            owner: connectedAddress,
+            image: metadata.image,
+            name: metadata.name,
+          });
+        } catch (e) {
+          notification.error("Error fetching your NTFs");
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/MyNfts.tsx:57]: "your NFTs"

File: extension/packages/nextjs/app/erc721/components/MyNfts.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    return (
+      <div className="flex justify-center items-center mt-10">
+        <span className="loading loading-spinner loading-lg"></span>
+      </div>
+    );
+
+  return (
+    <>
+      <div className="flex justify-center items-center space-x-2 flex-col sm:flex-row">
+        <p className="y-2 mr-2 font-bold text-2xl">Your Balance:</p>
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/components/MyNfts.tsx:81]: my-2?

File: extension/packages/nextjs/app/erc721/page.tsx [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            <p>
+              The ERC-721 Token Standard introduces a standard for Non-Fungible Tokens (
+              <a
+                target="_blank"
+                href="https://eips.ethereum.org/EIPS/eip-721"
+                className="underline font-bold text-nowrap"
+              >
+                EIP-721
+              </a>
+              ), in other words, each Token is unique.
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/page.tsx:37]: token

File: extension/packages/nextjs/app/erc721/page.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+              The ERC-721 token implementation uses the{" "}
+              <a
+                target="_blank"
+                href="https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/extensions/ERC721Enumerable.sol"
+                className="underline font-bold text-nowrap"
+              >
+                ERC-721 Enumerable extension
+              </a>{" "}
+              from OpenZeppelin to list all tokens and all the tokens owned by an address. You can remove this if you
+              plan to use an indexer, like a Subgraph or Ponder (extensions available).
```
Comment (Pabl0cks) [inline extension/packages/nextjs/app/erc721/page.tsx:60]: link to https://scaffoldeth.io/extensions

File: extension/README.md.args.mjs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+export const extraContents = `## 🚀 Setup ERC-721 NFT Extension
+
+This extension introduces an ERC-721 token contract and demonstrates how to use it, including getting the total supply and holder balance, listing all NFTs and NFTs from an address, and how to transfer.
```
Comment (Pabl0cks) [inline extension/README.md.args.mjs:3]: all NFTs from the contract / all NFTs from the collection

File: extension/README.md.args.mjs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+The ERC-721 Token Standard introduces a standard for Non-Fungible Tokens ([EIP-721](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol)), in other words, each Token is unique.
```
Comment (Pabl0cks) [inline extension/README.md.args.mjs:5]: token

**Review by Pabl0cks — COMMENTED** [review]
gj Damu! Extension is working nicely, just added some comments and nitpicks, almost all of them are just copys, messages or variable names.

File: extension/packages/nextjs/app/erc721/page.tsx [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+          </div>
+
+          <div className="divider my-0" />
+
+          <h2 className="text-3xl font-bold mt-4">Interact with the NFT</h2>
+
+          <div>
+            <p>Below you can mint an NFT for yourself or to another address.</p>
+            <p>
+              You can see your balance and your NFTs and after that you can see the total supply and all the NFTs
```
[the remark this answers is not in the document]
Comment (Pabl0cks) [reply extension/packages/nextjs/app/erc721/page.tsx:71]: My proposal is not the best probably, maybe you can just add some commas to yours 🙏

**Comment by damianmarti** [comment]
> gj Damu! Extension is working nicely, just added some comments and nitpicks, almost all of them are just copys, messages or variable names.

Thanks!!

**Review by Pabl0cks — APPROVED** [review]
Sorry Damu totally missed your updates until this week. All updates look good to me 🙌

Only pushed some tweaks on the Readme/description from ERC721 page, to make them all consistant.

Feel totally free to undo any changes or tweak them however you see them better! Was just quicker to push than report :)

**Comment by damianmarti** [comment]
> Sorry Damu totally missed your updates until this week. All updates look good to me 🙌
> 
> Only pushed some tweaks on the Readme/description from ERC721 page, to make them all consistant.
> 
> Feel totally free to undo any changes or tweak them however you see them better! Was just quicker to push than report :)

Thanks @Pabl0cks for the review and the tweaks!!

**Comment by damianmarti** [comment]
Merged this. I will add the extension to the extension's source of truth.

**Pull request merged**